### PR TITLE
Ensure ComponentBase async lifecycle methods return non-null task (imported from Blazor PR 1620)

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Imported from https://github.com/aspnet/Blazor/pull/1620

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
@rynowak Another one I've already reviewed/updated and consider ready, but need another approval before it can be merged. Please feel free to review further if you want!

File: src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs
Comment (rynowak) [inline src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs]: This is a nice imrovment and should fix some issues we were seeing with the execution context.

File: src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs
Comment (rynowak) [inline src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs]: There's an inconsistency here. A synchronously cancelled cancelled task will have its exception ignored - since `Task.IsCompleted` includes all three terminal states. https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.iscompleted?view=netframework-4.7.2

But an asynchronously cancelled task will have its exception reported.

IMO it would be good to skip reporting exceptions when a task is cancelled. Cancellation is usually the result of a shutdown or dispose in UI apps and you don't really need to know about it.

Comment (rynowak) [reply src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs]: It might be slightly more clear to implement the synchronous logic as a switch over https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskstatus?view=netframework-4.7.2

Comment (SteveSandersonMS) [reply src/Components/src/Microsoft.AspNetCore.Components/ComponentBase.cs]: Very good point. Done.

File: src/Components/test/Microsoft.AspNetCore.Components.Test/ComponentBaseTest.cs
Comment (SteveSandersonMS) [reply src/Components/test/Microsoft.AspNetCore.Components.Test/ComponentBaseTest.cs]: I've added some tests for that, but they are pretty limited right now. They would still be passing even if we *were* reporting the cancellation as exceptions, since the tests don't have any way to observe async exceptions from lifecycle methods.

The underlying limitation is that we don't have proper error reporting for that. I've filed a separate issue to track dealing with this: https://github.com/aspnet/AspNetCore/issues/4964. When that gets implemented, the unit tests I've added in this PR will only continue to pass if the task cancellation handling logic is correct (which I think it is anyway).

Comment (rynowak) [reply src/Components/test/Microsoft.AspNetCore.Components.Test/ComponentBaseTest.cs]: Yeah I think that's good enough for now.

**Review by rynowak — APPROVED** [review]

**Pull request merged**
